### DISCUSSION TITLE
How to create a Single Voice Dump (155 byte) from a 32 voices sysex file.

### DISCUSSION THREAD

**Discussion by safran5020** [asb2m10/dexed#444] [Q&A]
Hi Pascal,
First I want to say thanks for the great Dexed plugin.

I am an owner of a DX7 MK1 bought in 1985. I am also a programmer and I recently started scripting in Reaper in lua and more recently using ReaImgui. As a little project, I created a nice little DX7 Librarian (browser). So far I can display in a treeview all my sysex folders/files and upon clicking any sysex file I can display a list of its 32 voices.

I was hoping to be able to split any single voice out of the 4104 bytes file as each voice is 128 bytes. The problem is that a single voice dump is 155 bytes + 6 bytes for the header and 2 at the end for a total 163 bytes. For the header and the 2 at the end it's not a problem but how do I convert a 128 bytes to 155 bytes to build my string in order to send a single voice to my DX7 ?

I have looked at your source code and didn't see anywhere how you do this in the CART window when we right click a voice and select "Send program xyz to DX7".

I would be very grateful if you could lead me in the right direction.

Thanks !

En passant Pascal, je suis aussi Québecois, je suis de la grande région de Lanaudière. Le monde est petit !!

Merci

**Comment by FulopNandor** [comment]
Hi @safran5020!

If I interpret your question and the Dexed source code correctly, there are the following two methods in ``Source/PluginData.cpp`` you are looking for:
- [void Cartridge::unpackProgram(uint8_t *unpackPgm, int idx)](https://github.com/asb2m10/dexed/blob/2a2f8801abb714f64819970404c36671e782d4ae/Source/PluginData.cpp#L114): this performs the conversion of a single 128-byte compressed voice parameter array to the proper 155-byte single voice format parameter array,
- [void Cartridge::packProgram(uint8_t *src, int idx, String name, char *opSwitch)](https://github.com/asb2m10/dexed/blob/2a2f8801abb714f64819970404c36671e782d4ae/Source/PluginData.cpp#L56): this is the reverse, it produces the appropriate 128-byte compressed parameter array from a single 155-byte single voice format parameter array.

I hope I haven’t misled you too much with the above answer, but it might help until Pascal’s answer arrives.

**Comment by safran5020** [comment]
@FulopNandor ,

Thanks man for the quick reply. I guess you nailed it right on !!
I am glad my request wasn't to confusing, english is not my native language.

The code you highlighted seems to be exactly what I was looking for.
As soon as I have a chance I will look at it more closely, translate it in lua and give it a try in my script.

I will report back.

Big thanks again !

**Comment by asb2m10** [comment]
Thanks @FulopNandor for the quick answer; I think this is what @safran5020 was looking for.

@safran5020, take a look at this project https://dxconvert.martintarenskeen.nl/, there is some python scripts that might be what you are looking for also.

Oh tu es dans quel coin @safran5020 dans Lanaudière ? oui le monde est petit :)

**Comment by mtarenskeen** [comment]
First source of information: The original Yamaha DX7 MIDI SysEx documentation. VCED is the dataformat for single patchdumps, VMEM is the compressed dataformat that is used in the 32-patch banks. For the the DX7 MkII the ACED and AMEM dataformats were added to this.
